### PR TITLE
Make -m -4 work on xmonad

### PR DESCRIPTION
**Pull request by ironyman**
I hit this issue, https://github.com/DaveDavenport/rofi/issues/364.
When run with `-m -4`, rofi appears in the middle of my two displays (desktops in X I guess).

The problem is that `xcb_get_geometry` returns the coordinates relative to the root of the screen and not relative to viewport of the desktop. So the translation after that moves `t` outsides of the screen, `monitor_dimensions` cannot find any monitor `workarea` containing `t` and returns the whole display as `workarea`.

Another problem arises when there is no active window, in which case `xcb_get_geometry` triggers a segfault when `active_window == 0`.

I tested this on my setup on xmonad. I would like feedback from other people with similar setups.

```
Monitor layout:
              ID: 0
            name: eDP-1
        position: 0,0
            size: 2560,1440

              ID: 1 (primary)
            name: HDMI-2
        position: 2560,0
            size: 2560,1440


Compile time options:
        * window  enabled
        * drun    enabled
        * timings disabled
        * gcov    disabled
        * asan    disabled

For more information see: man rofi
                 Version: 1.3.1
              Bugreports: https://github.com/DaveDavenport/rofi/
                 Support: https://forum.qtools.org/
                          #rofi @ freenode.net
      Configuration file: /home/cyl/.config/rofi/config
```

### CODE REVIEW

**Comment by DaveDavenport** [comment]
Just out of curiosity I tried to reproduce the crash here, and it did not crash.
Do you have a backtrace? because the xcb_get_geometry_reply returns NULL and the code falls through.

Also here it does not position on the middle of the 2 monitors. Wonder why the difference.
The current translation should  give the right result as it gets the size of the monitor at set location.
The patch actually fails, as r->x and r->y is here the position inside the border of the window. (basically the border width)

File: source/xcb.c [hunk 1]
```diff
@@ -528,10 +528,25 @@ static int monitor_active_from_id ( int mon_id, workarea *mon )
...
+                    return TRUE;
+                }
+                // This is our give up point.
+                return FALSE;
+            }
             // get geometry.
             xcb_get_geometry_cookie_t c  = xcb_get_geometry ( xcb->connection, active_window );
             xcb_get_geometry_reply_t  *r = xcb_get_geometry_reply ( xcb->connection, c, NULL );
             if ( r ) {
+                // We don't need to translate.
```
Comment (DaveDavenport) [inline source/xcb.c]: We need translation, in I3 parent window is used to draw border, r->x, and r->y are relative to this window, not the root window.
Under my I3 this is equal to the border width.

Comment (ironyman) [reply source/xcb.c]: I added some print statements, https://github.com/ironyman/rofi/compare/next...ironyman:debugstuff

```
monitor_active_from_id -4
active window 14680074
active window geometry r 2565 37
translated geometry t 5131 75
mon 0 0 5120 1440
monitor_active_from_id -4
active window 14680074
active window geometry r 2565 37
translated geometry t 5131 75
mon 0 0 5120 1440
```

Note the `translated geometry t 5131 75`. 

I'm not sure what's supposed to happen in this case. What do you think?

Comment (ironyman) [reply source/xcb.c]: To give more context, my setup is two monitors side by side, each has width 2560. The active window is in the right one, hence the `2565 37`.

Comment (DaveDavenport) [reply source/xcb.c]: This is odd. I have almost similar setup but 1 full hd other 1600x1200.

For me it returns 4x4 for window geometry (the i3 border is 4 pixels). Translated is then right.
I should try to setup xmonad.

Comment (DaveDavenport) [reply source/xcb.c]:
```
monitor_active_from_id -4
active window 67108865
active window geometry r 2 2
translated geometry t 1939 43
mon 1920 0 1600 1200
monitor_active_from_id -4
active window 67108865
active window geometry r 2 2
translated geometry t 1939 43
mon 1920 0 1600 1200
```

**Comment by DaveDavenport** [comment]
Question, what type of setup do you have? xrandr? xinerama? something more exotic? zaphod?

**Comment by ironyman** [comment]
I'm using xrandr.

**Comment by cprussin** [comment]
FWIW I have a very similar setup (xrandr, xmonad, -monitor -4).  I don't experience the crash, but I _do_ experience the problem mentioned where rofi spans both monitors.  This PR resolved the issue for me.  Thanks!

**Comment by DaveDavenport** [comment]
When I have time I will install xmonad and try.  The patch seems to break things for !xmonad, so I cannot merge it.

**Comment by DaveDavenport** [comment]
FOR PACKAGERS: PLEASE DO NOT ADD THIS PATCH TO A PACKAGE.

I need some time to propperly investigate and come up with a fix that works everywhere (tm).

**Comment by ironyman** [comment]
Hey @DaveDavenport, I installed i3 and looked at the problem. The problem was that you were calling `xcb_translate_coordinates` with `active_window` as `src_window` instead of parent of `active_window`

**Comment by DaveDavenport** [comment]
Great!.

File: source/xcb.c [hunk 2]
```diff
@@ -528,11 +528,19 @@ static int monitor_active_from_id ( int mon_id, workarea *mon )
         xcb_get_property_cookie_t awc;
         awc = xcb_ewmh_get_active_window ( &xcb->ewmh, xcb->screen_nbr );
         if ( xcb_ewmh_get_active_window_reply ( &xcb->ewmh, awc, &active_window, NULL ) ) {
+            xcb_query_tree_cookie_t tree_cookie = xcb_query_tree ( xcb->connection, active_window );
+            xcb_query_tree_reply_t *tree_reply = xcb_query_tree_reply ( xcb->connection, tree_cookie, NULL );
```
Comment (DaveDavenport) [inline source/xcb.c]: Where is tree_reply free-ed.

File: source/xcb.c [hunk 3]
```diff
@@ -528,11 +528,19 @@ static int monitor_active_from_id ( int mon_id, workarea *mon )
...
+
+            if (!tree_reply) {
+                g_debug ( "Failed to get parent window, falling back to mouse location (-5)." );
+                goto fail;
```
Comment (DaveDavenport) [inline source/xcb.c]: Please no goto's

File: source/xcb.c [hunk 4]
```diff
@@ -528,39 +528,48 @@ static int monitor_active_from_id ( int mon_id, workarea *mon )
...
-                        free ( r );
-                        free ( t );
-                        return TRUE;
+            xcb_query_tree_cookie_t tree_cookie = xcb_query_tree ( xcb->connection, active_window );
+            xcb_query_tree_reply_t *tree_reply = xcb_query_tree_reply ( xcb->connection, tree_cookie, NULL );
+            if ( tree_reply ) {
+                // get geometry.
+                xcb_get_geometry_cookie_t c  = xcb_get_geometry ( xcb->connection, active_window );
+                xcb_get_geometry_reply_t  *r = xcb_get_geometry_reply ( xcb->connection, c, NULL );
+                if ( r ) {
```
Comment (DaveDavenport) [inline source/xcb.c]: r == NULL then tree_reply is never free 'ed.

**Comment by DaveDavenport** [comment]
Maybe it is good to split this one out in a subfunction:
(not well tested).
```patch
diff --git a/source/xcb.c b/source/xcb.c
index fbd3cd4..1473e58 100644
--- a/source/xcb.c
+++ b/source/xcb.c
@@ -481,6 +481,59 @@ static int monitor_active_from_winid ( xcb_drawable_t id, workarea *mon )
     }
     return FALSE;
 }
+
+static int monitor_active_from_id_focused ( int mon_id, workarea *mon )
+{
+    int retv = FALSE;
+    xcb_window_t              active_window;
+    xcb_get_property_cookie_t awc;
+    awc = xcb_ewmh_get_active_window ( &xcb->ewmh, xcb->screen_nbr );
+    if ( ! xcb_ewmh_get_active_window_reply ( &xcb->ewmh, awc, &active_window, NULL ) ) {
+        g_debug ( "Failed to get active window, falling back to mouse location (-5)." );
+        return retv;
+    }
+    xcb_query_tree_cookie_t tree_cookie = xcb_query_tree ( xcb->connection, active_window );
+    xcb_query_tree_reply_t *tree_reply = xcb_query_tree_reply ( xcb->connection, tree_cookie, NULL );
+    if ( !tree_reply ) {
+        g_debug ( "Failed to get parent window, falling back to mouse location (-5)." );
+        return retv;
+    }
+    // get geometry.
+    xcb_get_geometry_cookie_t c  = xcb_get_geometry ( xcb->connection, active_window );
+    xcb_get_geometry_reply_t  *r = xcb_get_geometry_reply ( xcb->connection, c, NULL );
+    if ( !r ) {
+        g_debug ( "Failed to get geometry of active window, falling back to mouse location (-5)." );
+        free ( tree_reply );
+        return retv;
+    }
+    xcb_translate_coordinates_cookie_t ct = xcb_translate_coordinates ( xcb->connection, tree_reply->parent, r->root, r->x, r->y );
+    xcb_translate_coordinates_reply_t  *t = xcb_translate_coordinates_reply ( xcb->connection, ct, NULL );
+    if ( t ) {
+        if ( mon_id == -2 ) {
+            // place the menu above the window
+            // if some window is focused, place menu above window, else fall
+            // back to selected monitor.
+            mon->x = t->dst_x - r->x;
+            mon->y = t->dst_y - r->y;
+            mon->w = r->width;
+            mon->h = r->height;
+            retv = TRUE;
+        }
+        else if ( mon_id == -4 ) {
+            monitor_dimensions ( t->dst_x, t->dst_y, mon );
+            retv = TRUE;
+        }
+        free ( t );
+    }
+    else {
+        g_debug ( "Failed to get translate position of active window, falling back to mouse location (-5)." );
+    }
+    free ( r );
+    free ( tree_reply );
+    return FALSE;
+}
+
+
 static int monitor_active_from_id ( int mon_id, workarea *mon )
 {
     xcb_window_t root = xcb->screen->root;
@@ -524,56 +577,8 @@ static int monitor_active_from_id ( int mon_id, workarea *mon )
         }
     }
     else if ( mon_id == -2 || mon_id == -4 ) {
-        xcb_window_t              active_window;
-        xcb_get_property_cookie_t awc;
-        awc = xcb_ewmh_get_active_window ( &xcb->ewmh, xcb->screen_nbr );
-        if ( xcb_ewmh_get_active_window_reply ( &xcb->ewmh, awc, &active_window, NULL ) ) {
-            xcb_query_tree_cookie_t tree_cookie = xcb_query_tree ( xcb->connection, active_window );
-            xcb_query_tree_reply_t *tree_reply = xcb_query_tree_reply ( xcb->connection, tree_cookie, NULL );
-            if ( tree_reply ) {
-                // get geometry.
-                xcb_get_geometry_cookie_t c  = xcb_get_geometry ( xcb->connection, active_window );
-                xcb_get_geometry_reply_t  *r = xcb_get_geometry_reply ( xcb->connection, c, NULL );
-                if ( r ) {
-                    xcb_translate_coordinates_cookie_t ct = xcb_translate_coordinates ( xcb->connection, tree_reply->parent, root, r->x, r->y );
-                    xcb_translate_coordinates_reply_t  *t = xcb_translate_coordinates_reply ( xcb->connection, ct, NULL );
-                    free ( tree_reply );
-                    if ( t ) {
-                        if ( mon_id == -2 ) {
-                            // place the menu above the window
-                            // if some window is focused, place menu above window, else fall
-                            // back to selected monitor.
-                            mon->x = t->dst_x - r->x;
-                            mon->y = t->dst_y - r->y;
-                            mon->w = r->width;
-                            mon->h = r->height;
-                            free ( r );
-                            free ( t );
-                            return TRUE;
-                        }
-                        else if ( mon_id == -4 ) {
-                            monitor_dimensions ( t->dst_x, t->dst_y, mon );
-                            free ( r );
-                            free ( t );
-                            return TRUE;
-                        }
-                    }
-                    else {
-                        g_debug ( "Failed to get translate position of active window, falling back to mouse location (-5)." );
-                    }
-                    free ( r );
-                }
-                else {
-                    g_debug ( "Failed to get geometry of active window, falling back to mouse location (-5)." );
-                }
-            }
-            else {
-                free ( tree_reply );
-                g_debug ( "Failed to get parent window, falling back to mouse location (-5)." );
-            }
-        }
-        else {
-            g_debug ( "Failed to get active window, falling back to mouse location (-5)." );
+        if ( monitor_active_from_id_focused ( mon_id, mon) ){
+            return TRUE;
         }
     }
     // Monitor that has mouse pointer.
```

**Comment by ironyman** [comment]
Tested on xmonad and i3, looks good.

**Comment by cprussin** [comment]
@DaveDavenport just curious, are you planning on a version bump to get a release out with this?  Thanks!

**Comment by cprussin** [comment]
@DaveDavenport just a gentle ping, any plans to do a version bump including this any time soon?  I'd love to switch back to a supported upstream package instead of a custom build...

**Comment by DaveDavenport** [comment]
yeah, there is one planned for asap.. I am just very busy. release notes are mostly done :D

**Comment by cprussin** [comment]
Wooo!  Thanks and sorry to bug you!

**Pull request merged**